### PR TITLE
Add all the typeclasses instance for the `zio.prelude.data.Optional` type (copied from the `Option` instances)

### PR DESCRIPTION
**Pull request by guizmaii**
@sideeffffect Did I forget any typeclass? 🤔

### CODE REVIEW

File: core/shared/src/main/scala/zio/prelude/data/Optional.scala [hunk 1]
```diff
@@ -119,7 +119,7 @@ sealed trait Optional[+A] { self =>
       case Optional.Absent       => Optional.Absent
     }
 
-  final def orElse[B >: A](other: Optional[B]): Optional[B] =
+  final def orElse[B >: A](other: => Optional[B]): Optional[B] =
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/data/Optional.scala:122]: The `Option#orElse` is lazy

File: core/shared/src/main/scala/zio/prelude/AssociativeBoth.scala [hunk 2]
```diff
@@ -1185,9 +1186,31 @@ object AssociativeBoth extends AssociativeBothLowPriority {
...
-          case _                  => None
+        fa match {
+          case Some(a) =>
+            fb match {
+              case Some(b) => Some((a, b))
+              case None    => None
+            }
+          case None    => None
+        }
+    }
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/AssociativeBoth.scala:1197]: Just optimizing the code here

File: core/shared/src/main/scala/zio/prelude/data/Optional.scala [hunk 3]
```diff
@@ -119,11 +119,8 @@ sealed trait Optional[+A] { self =>
       case Optional.Absent       => Optional.Absent
     }
 
-  final def orElse[B >: A](other: Optional[B]): Optional[B] =
-    self match {
-      case Optional.Present(_) => self
-      case Optional.Absent     => other
-    }
+  final def orElse[B >: A](alternative: => Optional[B]): Optional[B] =
+    if (isEmpty) alternative else this
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/data/Optional.scala:123]: Code copied from `Option#orElse`. Avoids the call to the `unapply` function in `Optional.Present(_)`

File: core/shared/src/main/scala/zio/prelude/data/Optional.scala [hunk 4]
```diff
@@ -119,11 +119,8 @@ sealed trait Optional[+A] { self =>
       case Optional.Absent       => Optional.Absent
     }
 
-  final def orElse[B >: A](other: Optional[B]): Optional[B] =
-    self match {
-      case Optional.Present(_) => self
-      case Optional.Absent     => other
-    }
+  final def orElse[B >: A](alternative: => Optional[B]): Optional[B] =
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/data/Optional.scala:122]: The `Option#orElse` method is lazy

File: core/shared/src/main/scala/zio/prelude/data/Optional.scala [hunk 5]
```diff
@@ -50,7 +50,7 @@ sealed trait Optional[+A] { self =>
       case Optional.Absent       => ifEmpty
     }
 
-  final def flatten[B](implicit ev: A <:< Option[B]): Option[B] =
+  final def flatten[B](implicit ev: A <:< Optional[B]): Optional[B] =
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/data/Optional.scala:53]: I didn't understand why the evidence what proving that `A` was an `Option[B]` nor why this method was returning an `Option[B]` and not an `Optional[B]`.
Probably a copy/paste mistake

File: core/shared/src/main/scala/zio/prelude/Equal.scala [hunk 6]
```diff
@@ -353,6 +354,16 @@ object Equal extends EqualVersionSpecific {
       case _                    => false
     }
 
+  /**
+   * Derives an `Equal[Optional[A]]` given an `Equal[A]`.
+   */
+  implicit def OptionalEqual[A: Equal]: Equal[Optional[A]] =
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/Equal.scala:360]: @sideeffffect I didn't find the tests for this typeclass 🤔

File: core/shared/src/main/scala/zio/prelude/Derive.scala [hunk 7]
```diff
@@ -100,6 +101,15 @@ object Derive {
         Equal.OptionEqual
     }
 
+  /**
+   * The [[DeriveEqual]] instance for [[Optional]].
+   */
+  implicit val OptionalDeriveEqual: DeriveEqual[Optional] =
```
Comment (guizmaii) [inline core/shared/src/main/scala/zio/prelude/Derive.scala:107]: @sideeffffect I didn't find the tests for this typeclass 🤔

**Review by sideeffffect — APPROVED** [review]
Oh, wonderful. This looks good :heart: Good to merge from what I can see.

**Comment by guizmaii** [comment]
@sideeffffect Can we make a new release maybe, please?

**Comment by sideeffffect** [comment]
Most definitely! https://github.com/zio/zio-prelude/releases/tag/v1.0.0-RC34

**Comment by sideeffffect** [comment]
@jdegoes could we please add @guizmaii the privileges to make releases? :pray:  He's proven to be an outstandingly good co-maintainer.

**Comment by guizmaii** [comment]
@sideeffffect Thanks for the compliments :)

I think I already have these rights, I just don't feel legitimate to take these decisions by myself for now but if you tell me I can, then let's go :)

**Comment by sideeffffect** [comment]
Please, feel free take these decisions yourself 😉

**Pull request merged**